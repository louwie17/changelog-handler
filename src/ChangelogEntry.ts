import * as fs from 'fs';
import YAML from 'yaml';
import { getBranchName, getGithubPRNumber } from './git';

export interface EntryOptions {
	changelogEntry: string;
	force: boolean;
	mergeRequest: number;
	dryRun: boolean;
	gitUsername: string;
	type: string;
}

export class ChangelogEntry {
	private options: EntryOptions;

	constructor(options: EntryOptions) {
		this.options = options;
	}

	public async execute() {
		// assert_feature_brach!
		// assert_title! unless editor
		// assert_new_file!
		// assert_valid_type!
		// if (this.options.dry_run) {
		//   write
		//   amend_commit if options.amend
		// }
		// if (editor) {
		// system("#{editor} '#{file_path}'");
		// }
		const yamlStr = YAML.stringify(this.options);
		await this.write(yamlStr);
	}

	private async write(yamlStr: string) {
		const branchName = await getBranchName();
		console.log(branchName);
		const prNumber = await getGithubPRNumber();
		fs.writeFileSync(
			`./changelogs/unreleased/${branchName}.yaml`,
			yamlStr,
			'utf8'
		);
	}
}
