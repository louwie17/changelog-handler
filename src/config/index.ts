import { Config } from './Config';
import { readConfig } from './readConfig';

export async function getConfig(
	defaultConfig: Config,
	extConfigPath?: string
): Promise<Config> {
	const extConfig = await readConfig(extConfigPath);
	if (!extConfig) {
		return defaultConfig;
	}
	return {
		...defaultConfig,
		...extConfig.config,
	};
}
