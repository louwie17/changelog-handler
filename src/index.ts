export * from './Release';
export * from './ChangelogEntry';
