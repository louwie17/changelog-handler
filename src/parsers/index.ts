export * from './Parser';
export * from './YamlParser';
