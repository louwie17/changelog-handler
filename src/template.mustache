== {{{ title }}} ({{#singleChange}}1 change{{/singleChange}}{{^singleChange}}{{ count }} changes{{/singleChange}})

{{#entries}}
- {{{ title }}} {{#merge_request}}(#{{merge_request}}){{/merge_request}}
{{/entries}}

{{#nochanges}}
No changes.
{{/nochanges}}
