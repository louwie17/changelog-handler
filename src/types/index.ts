export * from './ChangelogData';
export * from './EntryOptions';
