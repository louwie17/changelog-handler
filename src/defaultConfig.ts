import { Config } from './config/Config';
import { YamlParser } from './parsers/YamlParser';

export const defaultConfig: Config = {
	rootDir: '',
	parserType: 'yml',
	changelogPaths: {
		unreleased: './changelogs/unreleased',
		release: 'CHANGELOG.md',
	},
	customParsers: {
		yml: YamlParser,
	},
	changelogIdentifier: /^\=\= [0-9]\.[0-9]\.[0-9].*/,
};
