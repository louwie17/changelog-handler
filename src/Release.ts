import { existsSync, readdirSync, unlinkSync, writeFileSync } from 'fs';
import * as path from 'path';
import Mustache from 'mustache';

import { getConfig } from './config';
import { Config } from './config/Config';
import { defaultConfig } from './defaultConfig';
import { Parser } from './parsers';
import { ChangelogData, ReleaseOptions } from './types';
import markupTemplate from './template.mustache';
import {
  addCherrypickChangelog,
  insertChangelog,
} from './transform/insertChangelog';

/*
 * - Put the mark up in the necessary spot in the changelog
 * - Delete the unreleased files
 *
 * - List all change logs entries and allow use to select specifics
 */
export class Release {
  private config: Config;
  private parser: Parser<ChangelogData>;

  constructor(private options: ReleaseOptions) {
    // nothing
  }

  public async release() {
    await this.initialize();
  }

  public async execute() {
    await this.initialize();

    const entries = this.getChangelogEntries();
    if (entries.length === 0) {
      console.log('No changelogs available.');
      return;
    }
    const markdown = this.toMarkdown(entries.map((e) => e.data));
    const success = this.writeChangelog(
      markdown,
      entries.map((e) => e.data.merge_request)
    );
    if (success && !this.options.dryRun) {
      this.removeChangelogEntries(entries.map((e) => e.path));

      console.log(
        `Successfully updated: ${this.config.changelogPaths.release}`
      );
    }
  }

  private async initialize() {
    this.config = await getConfig(defaultConfig, this.options.config);
    const ParserClass = this.config.customParsers[this.config.parserType];
    this.parser = new ParserClass();
  }

  private getChangelogEntries(): {
    path: string;
    data: ChangelogData;
  }[] {
    if (!this.config) {
      return [];
    }
    const files = readdirSync(this.config.changelogPaths.unreleased);
    const entries = files.map((file) => {
      const filePath = this.config.changelogPaths.unreleased + '/' + file;
      return {
        path: filePath,
        data: this.parser.read(filePath),
      };
    });
    if (this.options.prNumbers) {
      return entries.filter((entry) => {
        return this.options.prNumbers?.includes(entry.data.merge_request);
      });
    }
    return entries;
  }

  private removeChangelogEntries(entries: string[]): void {
    for (const file of entries) {
      try {
        unlinkSync(file);
      } catch (err) {
        console.error(err);
      }
    }
  }

  private toMarkdown(entries: ChangelogData[]) {
    let date = this.options.date;
    if (!date && !this.options.title) {
      const today = new Date();
      date = today.toLocaleDateString();
    }
    // const template = readFileSync(markupTemplate, 'utf8'); for custom template
    const rendered = Mustache.render(markupTemplate, {
      entries,
      title: this.options.title
        ? this.options.title
        : `${this.options.version} ${date}`,
      count: entries.length,
      singleChange: entries.length === 1,
    });
    return rendered;
  }

  private writeChangelog(newChangelog: string, prNumbers: string[]) {
    const changelogPath = path.resolve(
      process.cwd(),
      this.config.changelogPaths.release
    );
    this.assertChangelogFile(changelogPath);
    if (this.options.dryRun) {
      console.log(`Writing to ${changelogPath}: \n${newChangelog}`);
      return;
    }
    if (this.options.cherryPick) {
      return addCherrypickChangelog(
        changelogPath,
        this.config.changelogIdentifier,
        this.options.version,
        newChangelog,
        prNumbers
      );
    }
    return insertChangelog(
      changelogPath,
      this.config.changelogIdentifier,
      newChangelog
    );
  }

  private assertChangelogFile(changelogPath: string) {
    if (!existsSync(changelogPath)) {
      console.warn(
        `Could not find changelog file: ${this.config.changelogPaths.release}, creating one instead.`
      );
      writeFileSync(changelogPath, '');
    }
  }
}
