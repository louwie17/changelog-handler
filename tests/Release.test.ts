import * as fs from 'fs';
jest.mock('fs');

import { Release } from '../src/Release';
import { ReleaseOptions } from '../src/types';
import testChangelog from './testChangelog.md';

const parserReadMock = jest.fn();
jest.mock('../src/parsers/YamlParser', () => ({
	YamlParser: jest.fn().mockImplementation(() => {
		return {
			read: parserReadMock,
		};
	}),
}));

class MockDate extends Date {
	constructor() {
		super('2020-05-14T11:01:58.135Z');
	}
	toLocaleDateString() {
		return '3/14/2021';
	}
}

describe('Release', () => {
	beforeEach(() => {
		jest.clearAllMocks();
		(fs.existsSync as jest.Mock).mockReturnValue(false);
		(fs.accessSync as jest.Mock).mockReturnValue(true);
	});

	describe('with no entries', () => {
		it('should do nothing and return: No changelogs available if no changelogs', async () => {
			(fs.readdirSync as jest.Mock).mockImplementation(
				(command: string) => {
					return [];
				}
			);
			jest.spyOn(global.console, 'log').mockImplementationOnce(() => {
				// do nothing.
			});
			const entry = new Release({ version: '1.0.0' } as ReleaseOptions);
			await entry.execute();

			expect(console.log).toHaveBeenCalledWith(
				'No changelogs available.'
			);
			expect(fs.readdirSync).toBeCalledTimes(1);
		});
	});

	describe('with entries', () => {
		beforeEach(() => {
			// (fs.readdirSync as jest.Mock).mockClear();
			(fs.readdirSync as jest.Mock).mockImplementation(
				(command: string) => {
					return ['1234-item.yml', '2222-change.yml'];
				}
			);
			(fs.readFileSync as jest.Mock).mockClear();
			(fs.readFileSync as jest.Mock).mockImplementation(
				(command: string) => {
					return '';
				}
			);
			parserReadMock.mockImplementation(() => {
				return {
					title: 'test',
					merge_request: 1234,
				};
			});
		});

		it('should load changelogs and add to changelog file using template', async () => {
			const entry = new Release({ version: '1.0.0' } as ReleaseOptions);
			await entry.execute();

			expect(fs.readdirSync).toBeCalledTimes(1);
			const changelog = (fs.writeFileSync as jest.Mock).mock.calls[0][1];
			const today = new Date();
			expect(changelog).toMatch(
				'== 1.0.0 ' + today.toLocaleDateString() + ' (2 changes)'
			);
			expect(changelog).toMatch('- test (#1234)');
			expect(fs.writeFileSync).toHaveBeenCalledTimes(1);
			expect(fs.unlinkSync).toHaveBeenCalledTimes(2);
			expect((fs.unlinkSync as jest.Mock).mock.calls[0][0]).toMatch(
				'1234-item.yml'
			);
			expect((fs.unlinkSync as jest.Mock).mock.calls[1][0]).toMatch(
				'2222-change.yml'
			);
		});

		it('should not call writeFileSync and unlinkSync for dry run', async () => {
			jest.spyOn(global.console, 'log').mockImplementationOnce(() => {
				// do nothing.
			});
			const entry = new Release({
				version: '1.0.0',
				dryRun: true,
			} as ReleaseOptions);
			await entry.execute();

			expect(fs.readdirSync).toBeCalledTimes(1);

			const changelog = (global.console.log as jest.Mock).mock
				.calls[0][0];
			const today = new Date();
			expect(changelog).toMatch(
				'== 1.0.0 ' + today.toLocaleDateString() + ' (2 changes)'
			);
			expect(changelog).toMatch('- test (#1234)');
			expect(fs.writeFileSync).toHaveBeenCalledTimes(0);
			expect(fs.unlinkSync).toHaveBeenCalledTimes(0);
		});

		it('should correct add new changelog above last one', async () => {
			(fs.readFileSync as jest.Mock).mockClear();
			console.log(testChangelog);
			(fs.readFileSync as jest.Mock).mockImplementation(
				(command: string) => {
					return testChangelog;
				}
			);
			const originalDate = global.Date;
			// @ts-ignore
			global.Date = MockDate;
			const entry = new Release({
				version: '1.0.0',
			} as ReleaseOptions);
			await entry.execute();

			expect(fs.readdirSync).toBeCalledTimes(1);

			const changelog = (fs.writeFileSync as jest.Mock).mock.calls[0][1];
			expect(changelog).toMatch('== 1.0.0 3/14/2021 (2 changes)');
			expect(changelog).toMatchSnapshot();
			expect(fs.writeFileSync).toHaveBeenCalledTimes(1);
			expect(fs.unlinkSync).toHaveBeenCalledTimes(2);
			global.Date = originalDate;
		});
	});
});
