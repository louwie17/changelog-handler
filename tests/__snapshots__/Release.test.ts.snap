// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Release with entries should correct add new changelog above last one 1`] = `
"My test changelog entries

== 1.0.0 3/14/2021 (2 changes)

- test (#1234)
- test (#1234)

== 1.1.1 3/14/2021 (1 change)
  - changelog (#1234)

== 1.1.0 3/14/2021 (1 change)
  - changelog (#1234)"
`;
